Python/optimizer.c

```diff
@@ -182,6 +182,12 @@ _PyOptimizer_Optimize(
     if (err <= 0) {
         return err;
     }
+
+    if (++interp->executors_created >= JIT_CLEANUP_THRESHOLD) {
```

Comment:
This could behave quite badly with many executors.

Suppose we have a 1000 executors:
Even in the ideal case, we need to scan all 1000, every time we create 10 new ones.

But what worries me is that we could clear all executors far too easily.
Suppose the VM starts compiling some new region of code, resulting in 20 new executors.
After the first 10, we will mark all executors as cold. After the second 10 we will clear all executors.

I'm not sure what the solution is, but I think a ticker that is independent of the number of executors created would be better. Maybe the GC trigger, or something similar?

Reply:
In earlier runs of the benchmarks, there was [no impact to performance](https://github.com/python/cpython/pull/123402#issuecomment-2319188505), even in instances where I tried pretty aggressive thresholds. In discussing this with Brandt, it seemed cheap to re-create in cases where we might be overzealous in invalidation.

Reply:
Probably also worth mentioning that I did experiment with triggering invalidation on gc earlier but in chatting with Brandt, it seemed like better design to have this tied to executor creation (i.e. since the invalidation isn't really at all coupled to gc today).

I'm sure that @pablogsal probably has thoughts here too?

Reply:
> I'm sure that @pablogsal probably has thoughts here too?

The other option that I think it may be sensible it's to do it in the eval breaker

For context: I would still prefer not to tie this with the gc as in general we have been holding off coupling the gc with things for a long time because any bug in the GC is already very non trivial and hard to debug and we have learn that (1) controlling any side-effects that are not purely derived from the generic algorithm and (2) keeping the GC from having to know what's operating on has been key to keep complexity at bay.

Reply:
> I'm not sure what the solution is, but I think a ticker that is independent of the number of executors created would be better. Maybe the GC trigger, or something similar?

I think the current approach is probably good enough for now, and we can iterate on the heuristics. The potential quadratic behavior (and wipe-out-everything-ness) of the current approach is a bit scary, but is unlikely to seriously impact most programs. It can also be mitigated by a follow-up PR to change "every ten executors" to "20% more executors" or "20% more JIT memory" or some other ratio-based approach, like the GC uses.

I also think the GC is probably the wrong place for this. We *could* add a new dedicated eval-breaker bit, but that sort of feels like overkill. The creation of new executors is a known safe point where (a) we *expect* to spend a bit of time managing JIT stuff and (b) we know JIT memory pressure will increase by some fixed amount. I don't think it would be any better to just set an eval breaker bit here and leave our new JIT code almost immediately to perform this same cleanup.